src/qz/utils/FileUtilities.java

```diff
@@ -61,25 +75,177 @@
+     */
+    public static boolean isGoodExtension(Path path) {
+        String fileName = path.getFileName().toString();
+
+        //"foo.exe." is valid on windows, but is immediately changed to "foo.exe" by the os
+        //this is undocumented behavior, therefore, rather than trying to support it, we fail it.
+        if (SystemUtilities.isWindows() && fileName.endsWith(".")) { return false; }
+
+        String[] tokens = fileName.split("\\.(?=[^.]+$)");
         if (tokens.length == 2) {
             String extension = tokens[1];
             for(String bad : FileUtilities.badExtensions) {
                 if (bad.equalsIgnoreCase(extension)) {
-                    return true;
+                    return false;
                 }
             }
         }
+        return true;
+    }
 
+    /**
+     * Returns whether or not the given file or folder is white-listed for File IO
+     * Currently hard-coded to the QZ data directory or anything provided by qz-tray.properties
+     * e.g. %APPDATA%/qz/data or $HOME/.qz/data, etc
+     */
+    public static boolean isWhiteListed(Path path, boolean allowRootDir, boolean sandbox, Certificate cert) {
+        String commonName = cert.isTrusted()? escapeFileName(cert.getCommonName()):"UNTRUSTED";
+        if (whiteList == null) {
+            populateWhiteList();
+        }
+
+        Path cleanPath = path.normalize().toAbsolutePath();
+        for(Pair<Path,String> allowed : whiteList) {
+            if (cleanPath.startsWith(allowed.getKey())) {
+                if ("".equals(allowed.getValue()) || allowed.getValue().contains("|" + commonName + "|") && (allowRootDir || !cleanPath.equals(allowed.getKey()))) {
+                    return true;
+                } else if (allowed.getValue().contains("|sandbox|")) {
+                    Path p;
+                    if (sandbox) {
+                        p = Paths.get(allowed.getKey().toString(), Constants.SANDBOX_DIR, commonName);
+                    } else {
+                        p = Paths.get(allowed.getKey().toString(), Constants.NOT_SANDBOX_DIR);
+                    }
+                    if (cleanPath.startsWith(p) && (allowRootDir || !cleanPath.equals(p))) {
+                        return true;
+                    }
+                }
+            }
+        }
         return false;
     }
 
+    private static void populateWhiteList() {
+        whiteList = new ArrayList<>();
+        //default sandbox locations. More can be added through the properties file
+        whiteList.add(new Pair<>(Paths.get(SystemUtilities.getDataDirectory()), "|sandbox|"));
+        whiteList.add(new Pair<>(Paths.get(SystemUtilities.getSharedDataDirectory()), "|sandbox|"));
+
+        Properties props = PrintSocketServer.getTrayProperties();
+        if (props != null) {
+            StringBuilder propString = new StringBuilder(props.getProperty("file.whitelist"));
```

Comment:
This can throw a NPE and needs a guard.  Probably just a `props.getProperty("file.whitelist", "")` would suffice.